miner/src/pool/queue.rs

```diff
@@ -0,0 +1,433 @@
+		// there is no need to invalidate the cache because of timestamp.
+		// Timestamp only affects local `PendingTransactions` with `Condition::Timestamp`.
+		if self.has_local_pending && current_timestamp > self.current_timestamp + TIMESTAMP_CACHE {
+			return None;
+		}
+
+		// It's fine to return limited set even if `nonce_cap` is `None`.
+		// The worst thing that may happen is that some transactions won't get propagated in current round,
+		// but they are not really valid in current block anyway. We will propagate them in the next round.
+		// Also there is no way to have both `Some` with different numbers since it depends on the block number
+		// and a constant parameter in schedule (`nonce_cap_increment`)
+		if self.nonce_cap.is_none() && nonce_cap.is_some() {
+			return None;
+		}
+
+		Some(pending.clone())
+	}
+}
+
+/// Ethereum Transaction Queue
+///
+/// Responsible for:
+/// - verifying incoming transactions
+/// - maintaining a pool of verified transactions.
+/// - returning an iterator for transactions that are ready to be included in block (pending)
+#[derive(Debug)]
+pub struct TransactionQueue {
+	insertion_id: Arc<AtomicUsize>,
+	pool: RwLock<Pool>,
+	options: RwLock<verifier::Options>,
+	cached_pending: RwLock<CachedPending>,
+}
+
+impl TransactionQueue {
+	/// Create new queue with given pool limits and initial verification options.
+	pub fn new(limits: txpool::Options, verification_options: verifier::Options, strategy: PrioritizationStrategy) -> Self {
+		TransactionQueue {
+			insertion_id: Default::default(),
+			pool: RwLock::new(txpool::Pool::new(Default::default(), scoring::NonceAndGasPrice(strategy), limits)),
+			options: RwLock::new(verification_options),
+			cached_pending: RwLock::new(CachedPending::none()),
+		}
+	}
+
+	/// Update verification options
+	///
+	/// Some parameters of verification may vary in time (like block gas limit or minimal gas price).
+	pub fn set_verifier_options(&self, options: verifier::Options) {
+		*self.options.write() = options;
+	}
+
+	/// Import a set of transactions to the pool.
+	///
+	/// Given blockchain and state access (Client)
+	/// verifies and imports transactions to the pool.
+	pub fn import<C: client::Client>(
+		&self,
+		client: C,
+		transactions: Vec<verifier::Transaction>,
+	) -> Vec<Result<(), transaction::Error>> {
+		// Run verification
+		let _timer = ::trace_time::PerfTimer::new("queue::verifyAndImport");
+		let options = self.options.read().clone();
+
+		let verifier = verifier::Verifier::new(client, options, self.insertion_id.clone());
+		let results = transactions
+			.into_iter()
+			.map(|transaction| verifier.verify_transaction(transaction))
+			.map(|result| match result {
+				Ok(verified) => match self.pool.write().import(verified) {
+					Ok(_imported) => Ok(()),
+					Err(err) => Err(convert_error(err)),
+				},
+				Err(err) => Err(err),
+			})
+			.collect::<Vec<_>>();
+
+		if results.iter().any(|r| r.is_ok()) {
```

Comment:
Didn't get the idea, why `any` is used here instead off `all`?

Also, if `all` was actually intended, all of this could be simplified by using the following `collect` transformation: `Vec<Result<T, E>>` → `Result<Vec<T>, E>`

Reply:
Ah, I probably missed the whole method's point. If I understood correctly, we need to return vector of transaction import results, since transactions are usually independent. And we don't want to stop processing the import if some of them failed.